### PR TITLE
feat: identity schema validation on startup

### PR DESCRIPTION
**Pull request by Benehiko**
## Related issue(s)

Fixes https://github.com/ory/kratos/issues/701

Relies on https://github.com/ory/jsonschema/pull/6 to be merged

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by Benehiko** [comment]
I have a proposal for implementing this without refactoring a lot of code in kratos. 
Currently in Kratos we are getting the config file after it has been validated by ory/x which gives us access to the configuration values. We then later read the schema files from the config which is used to validate new identities etc.

The current problems introduced by validating the schemas on boot:
- it breaks current tests easily
- the validation should happen after a provider was created by ory/x to access the config file values
- the validation on the custom schemas should happen again after the provider detected a change

I propose adding some lines of code in ory/x to accept an option with a function to run a validation job like so:

```go
configx.WithCustomSchemaValidator(func(p *configx.Provider, k *koanf.Koanf) error {
	// we need to be able to skip the validation of the identity schema as well
	// if the config is skipped since it could contain mock identity schema urls
	// which won't be loadable.
	c := &Config{l: l, p: p}
		return c.validateIdentitySchemas()
}),
```

and in ory/x we implement it like so:

```go
func (p *Provider) validate(k *koanf.Koanf) error {
	if p.skipValidation {
		return nil
	}

	out, err := k.Marshal(json.Parser())
	if err != nil {
		return errors.WithStack(err)
	}
	if err := p.validator.Validate(bytes.NewReader(out)); err != nil {
		p.onValidationError(k, err)
		return err
	}

	if p.customSchemaValidator != nil {
		if err = p.customSchemaValidator(p, k); err != nil {
			return err
		}
	}

	return nil
}
```

File: embedx/embedx.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	return nil, fmt.Errorf("the specified schema type (%d) is not supported", int(schema))
+}
+
+func (s SchemaType) GetSchemaID() string {
+	return schemas[s].id
+}
+
+func (s SchemaID) ToString() string {
+	return string(s)
+}
```
Comment (zepatrik) [inline embedx/embedx.go:82]:
```suggestion
```
unused?

File: embedx/embedx.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	Config SchemaType = iota
+	IdentityMeta
+	IdentityExtension
+)
+
+/*var (
+	ConfigSchemaID            SchemaID
+	IdentityMetaSchemaID      SchemaID
+	IdentityExtensionSchemaID SchemaID
+)*/
```
Comment (zepatrik) [inline embedx/embedx.go:35]:
```suggestion

const (
	Config SchemaType = iota
	IdentityMeta
	IdentityExtension
)
```

File: embedx/embedx.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+func (s SchemaType) GetSchemaID() string {
+	return schemas[s].id
+}
+
+func (s SchemaID) ToString() string {
+	return string(s)
+}
+
+// AddSchemaResources adds the logging schema to the compiler.
```
Comment (zepatrik) [inline embedx/embedx.go:84]:
```suggestion
// AddSchemaResources adds the specified schemas including their dependencies to the compiler.
```

File: embedx/embedx.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+					dependencies: nil,
+				},
+			},
+		},
+		IdentityExtension: {
+			id:           gjson.Get(IdentityExtensionSchema, "$id").Str,
+			data:         IdentityExtensionSchema,
+			dependencies: nil,
+		},
+	}
```
Comment (zepatrik) [inline embedx/embedx.go:66]: Deduplicate? You can make this an init func as well instead, then it is only a local variable. It is equivalent for the compiler.

```suggestion
	identityExt = &Schema{
		id:           gjson.Get(IdentityExtensionSchema, "$id").Str,
		data:         IdentityExtensionSchema,
		dependencies: nil,
	}
	schemas = map[SchemaType]*Schema{
		Config: {
			id:           gjson.Get(ConfigSchema, "$id").Str,
			data:         ConfigSchema,
			dependencies: nil,
		},
		IdentityMeta: {
			id:   gjson.Get(IdentityMetaSchema, "$id").Str,
			data: IdentityMetaSchema,
			dependencies: []*Schema{identityExt},
		},
		IdentityExtension: identityExt,
	}
```

File: driver/config/stub/.identity.invalid.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
+  "$id": "ory://identity-other-schema",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "IdentityOtherSchema",
```
Comment (zepatrik) [inline driver/config/stub/.identity.invalid.json:4]:
```suggestion
  "title": "IdentityInvalidSchema",
```

File: embedx/embedx_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+	"github.com/ory/jsonschema/v3"
+)
+
+func TestAddSchemaResources(t *testing.T) {
+
+	t.Run("case=validate schema resources are loading schema id correctly", func(t *testing.T) {
+		c := jsonschema.NewCompiler()
+		assert.NoError(t, AddSchemaResources(c, Config, IdentityMeta, IdentityExtension))
+
```
Comment (zepatrik) [inline embedx/embedx_test.go:16]: Loading is not required in this test right?
```suggestion
```

File: embedx/embedx_test.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	})
+
+	t.Run("case=add multiple schema resources", func(t *testing.T) {
+		c := jsonschema.NewCompiler()
+		assert.NoError(t, AddSchemaResources(c, Config, IdentityMeta))
+
+		_, err := c.Compile(IdentityMeta.GetSchemaID())
+		assert.NoError(t, err)
+
+		_, err = c.Compile(Config.GetSchemaID())
```
Comment (zepatrik) [inline embedx/embedx_test.go:45]:
```suggestion
		_, err = c.Compile(Config.GetSchemaID())
		assert.NoError(t, err)
```

File: embedx/embedx_test.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+		_, err := c.Compile(Config.GetSchemaID())
+		assert.NoError(t, err)
+
+		_, err = c.Compile(IdentityExtension.GetSchemaID())
+		assert.NoError(t, err)
+
+		_, err = c.Compile(IdentityMeta.GetSchemaID())
+		assert.NoError(t, err)
+	})
```
Comment (zepatrik) [inline embedx/embedx_test.go:61]: Make this a table test?
```suggestion
    for _, tc := range []struct {
		desc            string
		add             []SchemaType
		expectExtraDeps []SchemaType
	}{
		{
			desc: "config schema",
			add:  []SchemaType{Config},
		},
		{
			desc:            "identity meta incl dependency",
			add:             []SchemaType{IdentityMeta},
			expectExtraDeps: []SchemaType{IdentityExtension},
		},
		{
			desc:            "multiple schemas",
			add:             []SchemaType{Config, IdentityMeta},
			expectExtraDeps: []SchemaType{IdentityExtension},
		},
		{
			desc: "double-add",
			add:  []SchemaType{IdentityExtension, IdentityMeta},
		},
	} {
		t.Run("case="+tc.desc, func(t *testing.T) {
			c := jsonschema.NewCompiler()
			assert.NoError(t, AddSchemaResources(c, tc.add...))

			for _, s := range append(tc.add, tc.expectExtraDeps...) {
				_, err := c.Compile(s.GetSchemaID())
				assert.NoError(t, err)
			}
		})
	}
```

File: schema/extension.go [hunk 9]
```diff
@@ -57,13 +55,14 @@ type (
 
 func NewExtensionRunner(meta ExtensionRunnerMetaSchema, runners ...Extension) (*ExtensionRunner, error) {
```
Comment (zepatrik) [inline schema/extension.go:56]: This is always `embedx.IdentityExtension`
```suggestion
func NewExtensionRunner(runners ...Extension) (*ExtensionRunner, error) {
```

File: schema/extension.go [hunk 10]
```diff
@@ -57,13 +55,14 @@ type (
...
 	var err error
-	schema, err := ExtensionSchemas.ReadFile(string(meta))
-	if err != nil {
+	r := new(ExtensionRunner)
+	c := jsonschema.NewCompiler()
+
+	if err = embedx.AddSchemaResources(c, embedx.SchemaType(meta)); err != nil {
```
Comment (zepatrik) [inline schema/extension.go:61]:
```suggestion
	if err = embedx.AddSchemaResources(c, embedx.IdentityExtension); err != nil {
```

File: schema/extension.go [hunk 11]
```diff
@@ -57,13 +55,14 @@ type (
...
 		return nil, errors.WithStack(err)
 	}
 
-	r := new(ExtensionRunner)
-	r.meta, err = jsonschema.CompileString(string(meta), string(schema))
+	r.meta, err = c.Compile(embedx.SchemaType(meta).GetSchemaID())
```
Comment (zepatrik) [inline schema/extension.go:65]:
```suggestion
	r.meta, err = c.Compile(embedx.IdentityExtension.GetSchemaID())
```

File: schema/extension.go [hunk 12]
```diff
@@ -2,24 +2,22 @@ package schema
...
 
 	"github.com/pkg/errors"
 
 	"github.com/ory/jsonschema/v3"
+	"github.com/ory/kratos/embedx"
 )
 
-//go:embed .schema/extension/*.json
-var ExtensionSchemas embed.FS
+var ExtensionRunnerIdentityMetaSchema = ExtensionRunnerMetaSchema(embedx.IdentityExtension)
```
Comment (zepatrik) [inline schema/extension.go:13]:
```suggestion
```

File: schema/extension.go [hunk 13]
```diff
@@ -2,24 +2,22 @@ package schema
...
 
 const (
-	ExtensionRunnerIdentityMetaSchema ExtensionRunnerMetaSchema = ".schema/extension/identity.schema.json"
-	extensionName                     string                    = "ory.sh/kratos"
+	extensionName string = "ory.sh/kratos"
 )
 
 type (
-	ExtensionRunnerMetaSchema string
+	ExtensionRunnerMetaSchema embedx.SchemaType
```
Comment (zepatrik) [inline schema/extension.go:20]:
```suggestion
```

**Review by zepatrik — CHANGES_REQUESTED** [review]
Please also update the render version schema CI parameter: https://github.com/ory/kratos/blob/3409eda60ab7095feffd51058b9040e62faccc5f/.circleci/config.yml#L223

File: driver/config/config.go [hunk 14]
```diff
@@ -235,22 +241,75 @@ func MustNew(t *testing.T, l *logrusx.Logger, opts ...configx.OptionModifier) *C
...
+func (p *Config) validateIdentitySchemas() error {
+	for _, s := range p.IdentityTraitsSchemas() {
+		resource, err := jsonschema.LoadURL(s.URL)
+		if err != nil {
+			return err
+		}
+		j, err := p.getIdentitySchemaValidator()
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (Benehiko) [reply driver/config/config.go:307]: ahh yes xD

File: embedx/embedx.go (hunk 1 above)
Comment (Benehiko) [reply embedx/embedx.go:82]: ahh yeah forgot to take this out. I refactored a lot of code here

**Comment by aeneasr** [comment]
This ready to be merged?

**Comment by Benehiko** [comment]
> This ready to be merged?

not yet, addressing @zepatrik's comments first

File: embedx/embedx_test.go (hunk 6 above)
Comment (Benehiko) [reply embedx/embedx_test.go]: yeah it wasn't. It used to be required before i refactored embedx.
I have changed it now

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
This test failure looks new: https://app.circleci.com/pipelines/github/ory/kratos/4967/workflows/358ca1e3-64fd-42cb-b76c-5d1a7b19b852/jobs/26571

**Comment by Benehiko** [comment]
> This test failure looks new: https://app.circleci.com/pipelines/github/ory/kratos/4967/workflows/358ca1e3-64fd-42cb-b76c-5d1a7b19b852/jobs/26571

This is a new test added in the last commit. It passes locally, not sure why it is failing on `test-coverage`. My other tests also randomly failed on it and then suddenly passed on a force push.

**Comment by aeneasr** [comment]
Execute `go run -count=100 [....]` to run 100 tests and find flakes. I recommend running short test so `go run -count=100 -short -tags sqlite ./...` e.g.

**Comment by Benehiko** [comment]
@aeneasr codecov/project fails with -0.02%
Can we still merge this?

**Comment by aeneasr** [comment]
yes :)

**Pull request merged**